runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+        return startAddress;
+    }
+
+    /** Append zero or more CONTINUATION records to this
+     *  object's stream.  Each will contain a fraction of
+     *  the state of the object that we're checkpointing
+     *  (up to batchSize items at a time).
+     *
+     *  The Iterators class appears to preserve the laziness
+     *  of Stream processing; we don't wish to use more
+     *  memory than strictly necessary to generate the
+     *  checkpoint.  NOTE: It would be even more useful if
+     *  the map had a lazy iterator: the eagerness of
+     *  map.keySet().stream() is not ideal, but at least
+     *  it should be much smaller than the entire map.
+     *
+     *  NOTE: The postAppendFunc lambda is executed in the
+     *  current thread context, i.e., inside of a Corfu
+     *  transaction, and that transaction will be *aborted*
+     *  at the end of this function.  Any Corfu data
+     *  modifying ops will be undone by the TXAbort().
+     *
+     *  TX side-effect: We call TXAbort().
+     *
+     * @return Stream of global log addresses of the
+     * CONTINUATION records written.
+     */
+    public List<Long> appendObjectState() {
+        ImmutableMap<String,String> mdKV = ImmutableMap.copyOf(this.mdKV);
+        List<Long> continuationAddresses = new ArrayList<>();
+
+        try {
+            Iterators.partition(map.keySet().stream()
+                    .map(k -> {
+                        return new SMREntry("put",
+                                new Object[]{keyMutator.apply(k), valueMutator.apply(map.get(k))},
+                                Serializers.JSON);
+                    }).iterator(), batchSize)
+                    .forEachRemaining(entries -> {
+                        // Convert Object[] to SMREntry[], combined with byte count.
+                        // java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Lorg.corfudb.protocols.logprotocol.SMREntry;
+                        int numEntries = ((List) entries).size();
+                        SMREntry e[] = new SMREntry[numEntries];
+                        for (int i = 0; i < numEntries; i++) {
+                            e[i] = (SMREntry) ((List) entries).get(i);
+                        }
+                        CheckpointEntry cp = new CheckpointEntry(CheckpointEntry.CheckpointEntryType.CONTINUATION,
+                                author, checkpointID, mdKV, e);
+                        long pos = sv.append(cp, null, null);
+                        postAppendFunc.accept(cp, pos);
+                        continuationAddresses.add(pos);
+                        numEntries++;
+                        // CheckpointEntry::serialize() has a side-effect we use
+                        // for an accurate count of serialized bytes of SRMEntries.
+                        numBytes += cp.getSmrEntriesBytes();
+                    });
+        } finally {
+            rt.getObjectsView().TXAbort();
```

Comment:
I will rearrange the transaction handling here.  The general absence of try/catch is intentional ... but I'll address further that in a PR-wide comment.